### PR TITLE
JA3 for java 

### PR DESCRIPTION
**Pull request by ajathan**

### CODE REVIEW

File: logback.xml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration debug="false" scan="true" scanPeriod="60 seconds">
+    <jmxConfigurator />
```
Comment (krdev) [inline logback.xml]: Why do you need your own version of logback config? talk to Vikram

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    /**
+     * Calculate JA3 string from a ClientHello packet.Note that we do not compute an MD5 hash here.
+     *
+     * @param packet packet to inspect
+     * @return JA3 fingerprint or null if no TLS ClientHello detected in given packet
+     * @see <a href="https://github.com/salesforce/ja3">Original JA3 implementation</a>
+     */
+    public static String ja3Signature(final ByteBuffer packet) {
```
Comment (krdev) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: Why static? we could have an instance in our app context (imap/pop) and pass it in where needed. Talk to vikram also.

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+     * https://tools.ietf.org/html/draft-davidben-tls-grease-01.
+     */
+    private static final int[] GREASE = new int[] { 0x0a0a, 0x1a1a, 0x2a2a, 0x3a3a, 0x4a4a, 0x5a5a, 0x6a6a, 0x7a7a, 0x8a8a, 0x9a9a, 0xaaaa, 0xbaba,
+            0xcaca, 0xdada, 0xeaea, 0xfafa };
+
+
+    /**
+     * Private constructor.
+     */
+    private JA3Signature() {
```
Comment (krdev) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: Is this dependent on the TLS version? Will you support any version or only 1.2 and above??

**Review by krdev — COMMENTED** [review]
Can you also copy over a version of .travis.yml and cd directory contents, copy these from jmetrics project

File: logback.xml (hunk 1 above)
Comment (ajathan) [reply logback.xml]: removed this file

**Comment by ajathan** [comment]
copied .travis.yml and cd directory contents from jmetrics

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java (hunk 2 above)
Comment (ajathan) [reply src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: spoke to Vikram. Wont be super useful adding this to the app context

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+            <artifactId>jsr305</artifactId>
+            <version>3.0.1</version>
+        </dependency>
+        <dependency>
+            <groupId>com.github.lafa.logfast</groupId>
+            <artifactId>logfast.core</artifactId>
+            <version>1.0.6</version>
+        </dependency>
+
+        <!-- dependency> <groupId>ch.qos.logback</groupId> <artifactId>logback-classic</artifactId>
```
Comment (nvsnvikram) [inline pom.xml]: remove

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>1.10.19</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.findbugs</groupId>
```
Comment (nvsnvikram) [inline pom.xml]: wouldn't this get packaged if scope is non test?

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+            <artifactId>jsr305</artifactId>
+            <version>3.0.1</version>
+        </dependency>
+        <dependency>
```
Comment (nvsnvikram) [inline pom.xml]: Where are we using logfast?

File: src/main/java/com/lafaspot/ja3_4java/package-info.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+/**
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/package-info.java:1]: Empty comments

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+	private static final int SIXTEEN = 16;
+	/**
+	 * Constant value for new line.
+	 */
+	private static final byte NEWLINE = 0x0a;
+	/**
+	 * Constant value for vertical tab.
+	 */
+	private static final byte VERTICALTAB = 0x0b;
+	/**
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java:47]: Empty comments

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+     */
+    private static final byte CLIENT_HELLO = 1;
+    /**
+     * Client hello random lenght.
+     */
+    private static final byte CLIENT_HELLO_RANDOM_LEN = 32;
+	/**
+	 * Constant for the value 4.
+	 */
+	private static final int FOUR = 4;
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: MIN_PACKET_LENGTH

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+	/**
+	 * Constant value 3.
+	 */
+	private static final int THREE = 3;
+	/**
+	 * Constant value 8.
+	 */
+	private static final int EIGHT = 8;
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: ONE_BYTE

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+	/**
+	 * Constant value 16.
+	 */
+	private static final int SIXTEEN = 16;
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: TWO_BYTES

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java (hunk 2 above)
Comment (ajathan) [reply src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: removed static

File: pom.xml (hunk 5 above)
Comment (ajathan) [reply pom.xml]: removed

File: pom.xml (hunk 6 above)
Comment (ajathan) [reply pom.xml]: removed

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java (hunk 8 above)
Comment (ajathan) [reply src/main/java/com/lafaspot/ja3_4java/JA3Signature.java:47]: resolved

File: src/main/java/com/lafaspot/ja3_4java/JA3Signature.java (hunk 11 above)
Comment (ajathan) [reply src/main/java/com/lafaspot/ja3_4java/JA3Signature.java]: done

**Review by nvsnvikram — APPROVED** [review]

**Review by nvsnvikram — APPROVED** [review]

**Pull request merged**